### PR TITLE
Gp/zend233_zend234

### PR DESCRIPTION
**Pull request by JackPiri**
PR bringing in minor refactoring.

### CODE REVIEW

File: src/init.cpp
Comment (ptagl) [inline src/init.cpp]: Due to the internal logic of the `metrics thread` (responsible for the UI) this reindexing message is never shown because it gets continuously overwritten by the `printStats()` function.

I wouldn't remove the `InitMessage()` here at the moment, but it might be the case to investigate in the future how we can improve the update logic of the UI so that all the important information is shown. For my personal preference, I would like to see the normal block height counter increasing and the "reindex" message as well, so that the user can clearly understand that the blocks are not being downloaded from other peers.

File: src/sc/proofverifier.cpp
Comment (ptagl) [inline src/sc/proofverifier.cpp:530]: Given that it's never possible to end up in a state in which the verifier returned `true` and the error code is different from `CctpErrorCode::OK`, what about adding an assert?

```suggestion
    assert(code == CctpErrorCode::OK);
    return ProofVerificationResult::Passed;
```

In case, let's do the same for the function `NormalVerifyCsw()`.

**Review by ptagl — CHANGES_REQUESTED** [review]

File: src/sc/proofverifier.cpp
Comment (JackPiri) [reply src/sc/proofverifier.cpp:530]: Makes sense, addressed in 222155ba23849d56ca62d84e830bb301e9c88186.

**Review by a-petrini — APPROVED** [review]

File: src/init.cpp
Comment (JackPiri) [reply src/init.cpp]: Addressed in 0d6cab92c1103b9fb2791a767c1599387ce3bff9.

File: src/metrics.cpp
Comment (a-petrini) [inline src/metrics.cpp:89]: Why creating another `boost::synchronized_value` object, and not using the same `initMessage`?
ThreadImport() sets its value sequentially (at first when doing header reindex, and then when doing a proper reindex). Outside the thread, `AppInit2()` also sets its value sequentially without the risk of being overwritten by other threads (`ThreadImport()` and `StartNode()`, which are started after the last access to `initMessage` by `AppInit2()`)

Comment (JackPiri) [reply src/metrics.cpp:89]: We have two different threads acting (in the sense of setting status and messages, a third thread running `ThreadShowMetricsScreen` is responsible of actually printing to GUI) here:

a->`AppInit2` (main thread)
1a) "Verifying wallet..."
2a) "Loading block index..."
3a) "Verifying blocks..."
4a) "Loading wallet..."
5a) "Activating best chain..."
5.5a) ThreadImport starts on another thread
6a) "Done loading" (via `SetDoneLoadingUI`)
7a) no printing if not importing, otherwise print importing info

b->`ThreadImport` (another thread)
1b) "Reindexing block headers from files...(blk%05u.dat)"
2b) "Reindexing block from files...(blk%05u.dat)"

Now, due to the synchronization mechanism:
```
while (!fRequestShutdown && chainActive.Tip() == NULL)
    MilliSleep(10);
```
the actual sequence of events is (considering for example 3 blkXXXXX.dat available):

...
4a)
5a)
1b) blk00000.dat, blk00001.dat, blk00002.dat
2b*) blk00000.dat
6a)
...) some minutes pass
2b**) blk00001.dat,
...) some minutes pass
2b***) blk00002.dat

Hence, if `initMessageAfterLoading` would not be present (having only `initMessage`):
1. the step 2b* would be immediately overwritten, with "Done Loading"
2. after printing "Done Loading" the GUI would not print `initMessage` (resulting in the user feeling all the loading has been completed and no importing is underway, guessing the possible blockchain progress due to p2p download)
3. after some minutes (this is the time required for importing a blkXXXXX.dat file; full importing, not headers only), step 2b** would trigger an update of `initMessage`

Conversely, introducing `initMessageAfterLoading`, the messages are stored in two different strings and the decision related to which string should be printed is based on the loading/importing status.

File: src/metrics.cpp
Comment (AndreaLanfranchi) [inline src/metrics.cpp]: std::endl means "append new line char and flush stream"
This should be written as 
```suggestion
    std::cout << msg << "\n" << std::endl;
```

File: src/init.cpp
Comment (AndreaLanfranchi) [inline src/init.cpp]: Why casting nFile to unsigned int when it's simpler to change the format argument from 05u to 05i ?
Also conversion to c_str() is pleonastic and causes a double conversion.

The operator `_` tries translate the string produced by tfm::format. The way it is used makes no sense as :
- we don't actually provide any translation
- the way it is written would impose we translate all possible strings resulting from all possible combinations of blkxxxxx.dat

If we really want to keep the translation it should be written as 
uiInterface.InitMessage(tfm::format(_("Reindexing block headers from files...") + "(blk%05u.dat)", (unsigned int)nFile));

File: src/init.cpp
Comment (AndreaLanfranchi) [inline src/init.cpp]: There is no point in initializing struct pos before we actually check for file existence.
Move this after the existence check

File: src/init.cpp
Comment (AndreaLanfranchi) [inline src/init.cpp]: This logic is broken.
If OpenBlockFile returns NULL (considering we've already done existence check - which only verifies the name of the file exists in file system but it's location has not been reached yet) it means we likely have an IO error. In such case the loop breaks as if it has traversed all the files and considers he task done which is not true.

**Review by AndreaLanfranchi — CHANGES_REQUESTED** [review]

File: src/init.cpp
Comment (JackPiri) [reply src/init.cpp]: Modified with:
`uiInterface.InitMessage(tfm::format(_("Reindexing block headers from files...") + "(blk%05i.dat)", nFile));`
and:
`uiInterface.InitMessageAfterLoading(tfm::format(_("Reindexing block from files...") + "(blk%05i.dat)", nFile));`.

File: src/metrics.cpp
Comment (JackPiri) [reply src/metrics.cpp]: Agree with your suggestion, modified accordingly.

File: src/init.cpp
Comment (JackPiri) [reply src/init.cpp]: Cannot move after because `pos `is required by `GetBlockPosFilename`.

File: src/init.cpp
Comment (JackPiri) [reply src/init.cpp]: Do you think adding a retry and if failing again printing an error message on screen (at log level this is already the case) can be ok?
In first place I opted for inserting an assert but maybe this is a little too severe; indeed if there is some error when importing, the node would keep the blockchain up to current importing level and proceed downloading the remaining blocks from p2p network.
Do you have any other proposal?

File: src/init.cpp
Comment (AndreaLanfranchi) [reply src/init.cpp]: What about an overload of function GetBlockPosFilename() which receives nFile instead of pos considering in that function the only relevant payload of pos is nfile ?

Comment (AndreaLanfranchi) [reply src/init.cpp]: Or even better ... Initialize only once outside the loop and for every iteration reset nfile and pos.
So we stop allocating and destructing on every loop.

File: src/init.cpp
Comment (AndreaLanfranchi) [reply src/init.cpp]: The problem here is that this thread is labelled "Importing" but actually serves the reindex purpose too.
If something goes south during reindex (or fast reindex) the data on disk is, as you say, to last importing level **but indexes** are inconsistent and we have wrongly stated that the reindex has completed correctly.
Which means all RPC calls which rely on correct indexes might return inconsistent data.

File: src/init.cpp
Comment (JackPiri) [reply src/init.cpp]: Yes, I have opted for constructing once outside loop and then resetting at the end of each iteration.

File: src/init.cpp
Comment (AndreaLanfranchi) [reply src/init.cpp]: I would say in this case an assert would be more appropriate

File: src/init.cpp
Comment (AndreaLanfranchi) [inline src/init.cpp]: Prefer pre increment rather than post increment

File: src/init.cpp
Comment (AndreaLanfranchi) [inline src/init.cpp]: Probably these two lines should stay in the `if (freindexFast)` branched code: in fact if that was false we never incremented pos and therefore no need to reset it to 0.0

**Review by AndreaLanfranchi — CHANGES_REQUESTED** [review]

File: src/metrics.cpp
Comment (drgora) [inline src/metrics.cpp]: This is unlikely, but now that this variable is detached from a specific string, we should make sure that we do not have unexpected memory issues with memory order. I'd suggest to make this an atomic variable:
```suggestion
std::atomic<bool> loaded = false;
```

File: src/init.cpp
Comment (AndreaLanfranchi) [reply src/init.cpp]: I don't agree with Paolo's analysis.
As shown in Jira ticket the message of reindexing was printed and stuck to the very same value till the reindexing finishes.
I believe several additions here introduced were not really necessary

Comment (AndreaLanfranchi) [reply src/init.cpp]: The message "Done loading" (which sets loaded == true in metrics thread) is not shown until chainActive.Tip() != NULL: and this condition happens only after reindex has completed (main thread stays in sleep wait every 10 milliseconds)

File: src/metrics.cpp
Comment (AndreaLanfranchi) [reply src/metrics.cpp:89]: I agree with @a-petrini : all additions are not needed here.
This
```
while (!fRequestShutdown && chainActive.Tip() == NULL)
    MilliSleep(10);
```
actually suspends main thread until all reindexing finishes i.e. after InitBlockIndex is completely finalized (which instead does nothing if either fReindex or fReindexFast are true)

File: src/metrics.cpp
Comment (JackPiri) [reply src/metrics.cpp]: Agree with your suggestion, modified accordingly.

**Review by drgora — APPROVED** [review]
This is ok for me, thanks.

File: src/init.cpp
Comment (JackPiri) [reply src/init.cpp]: Modified accordingly.

File: src/init.cpp
Comment (JackPiri) [reply src/init.cpp]: Modified accordingly.

File: src/init.cpp
Comment (JackPiri) [reply src/init.cpp]: Modified accordingly.

File: src/init.cpp
Comment (ptagl) [reply src/init.cpp]: Unfortunately `chainActive.Tip()` is NULL only during the processing of the headers, as soon as the first block is read from file and processed it's not NULL anymore, that's why we need such change.

**Review by ptagl — APPROVED** [review]

**Comment by ptagl** [comment]
Looks good to me, thanks!

@AndreaLanfranchi waiting for your approval to proceed with the merge.

**Review by AndreaLanfranchi — APPROVED** [review]

**Pull request merged**
